src/modules/fancyzones/lib/ZoneWindow.cpp

```diff
@@ -1,765 +1,774 @@
+    }
+}
+
+IFACEMETHODIMP_(void)
+ZoneWindow::CycleActiveZoneSet(DWORD wparam) noexcept
+{
+    CycleActiveZoneSetInternal(wparam, Trace::ZoneWindow::InputMode::Keyboard);
+
+    if (m_windowMoveSize)
+    {
+        InvalidateRect(m_window.get(), nullptr, true);
+    }
+    else
+    {
+        FlashZones();
+    }
+}
+
+IFACEMETHODIMP_(void)
+ZoneWindow::SaveWindowProcessToZoneIndex(HWND window) noexcept
+{
     auto processPath = get_process_path(window);
-    if (!processPath.empty())
+    if (!processPath.empty() && m_activeZoneSet)
     {
         DWORD zoneIndex = static_cast<DWORD>(m_activeZoneSet->GetZoneIndexFromWindow(window));
         if (zoneIndex != -1)
         {
-            RegistryHelpers::SaveAppLastZone(window, processPath.data(), zoneIndex);
+            JSONHelpers::FancyZonesDataInstance().SetAppLastZone(window, processPath.data(), zoneIndex);
+        }
+    }
+}
+
+#pragma region private
+void ZoneWindow::ShowZoneWindow() noexcept
+{
+    if (m_window)
+    {
+        m_flashMode = false;
+
+        UINT flags = SWP_NOSIZE | SWP_NOMOVE | SWP_NOACTIVATE;
+
+        HWND windowInsertAfter = m_windowMoveSize;
+        if (windowInsertAfter == nullptr)
+        {
+            windowInsertAfter = HWND_TOPMOST;
         }
-    }
+
+        SetWindowPos(m_window.get(), windowInsertAfter, 0, 0, 0, 0, flags);
+
+        AnimateWindow(m_window.get(), m_showAnimationDuration, AW_BLEND);
+        InvalidateRect(m_window.get(), nullptr, true);
+    }
+}
+
+void ZoneWindow::HideZoneWindow() noexcept
+{
+    if (m_window)
+    {
+        ShowWindow(m_window.get(), SW_HIDE);
+        m_keyLast = 0;
+        m_windowMoveSize = nullptr;
+        m_drawHints = false;
+        m_highlightZone = nullptr;
+    }
+}
+
+void ZoneWindow::LoadSettings() noexcept
+{
+    auto& deviceInfoMap = JSONHelpers::FancyZonesDataInstance().GetDeviceInfoMap();
+    if (!deviceInfoMap.contains(m_uniqueId))
+    {
+        // Creates entry in map when ZoneWindow is created
+        deviceInfoMap[m_uniqueId] = JSONHelpers::DeviceInfoData{ JSONHelpers::ZoneSetData{ L"null", JSONHelpers::ZoneSetLayoutType::Grid, 1 }, true, 16, 3 };
+
+        JSONHelpers::FancyZonesDataInstance().MigrateDeviceInfoFromRegistry(m_uniqueId);
+    }
+
+    JSONHelpers::FancyZonesDataInstance().ParseDeviceInfoFromTmpFile(ZoneWindowUtils::GetActiveZoneSetTmpPath());
+    JSONHelpers::FancyZonesDataInstance().ParseDeletedCustomZoneSetsFromTmpFile(ZoneWindowUtils::GetCustomZoneSetsTmpPath());
+}
+
+void ZoneWindow::InitializeZoneSets(MONITORINFO const& mi) noexcept
+{
+    CalculateZoneSet();
+
+    if (!m_activeZoneSet)
+    {
+        ChooseDefaultActiveZoneSet();
+    }
+}
+
+void ZoneWindow::CalculateZoneSet() noexcept
+{
+    const auto& deviceInfoMap = JSONHelpers::FancyZonesDataInstance().GetDeviceInfoMap();
+    const auto& activeDeviceId = JSONHelpers::FancyZonesDataInstance().GetActiveDeviceId();
+    const auto& activeZoneSet = deviceInfoMap.at(m_uniqueId).activeZoneSet;

```

Comment:
Are we sure that m_uniqueId is present in the map? `.at` throws, though the function is marked as `noexcept`, i.e. we'll crash.

Reply:
We are, because in function invoked before this one (`LoadSettings`) we check if device info map is empty for given unique id, and add `default` layout for it. So we will always have that entry in the map. We can add check of course, but I think it's not needed in this particular case.